model/waterfall.go

```diff
@@ -0,0 +1,618 @@
+		}
+	default:
+		return task.Status
+	}
+}
+
+// createWaterfallTasks takes ina  build's task cache returns a list of waterfallTasks.
+func createWaterfallTasks(tasks []build.TaskCache) ([]waterfallTask, task.TaskStatusCount) {
+	//initialize and set TaskStatusCount fields to zero
+	statusCount := task.TaskStatusCount{}
+	waterfallTasks := []waterfallTask{}
+
+	// add the tasks to the build
+	for _, t := range tasks {
+		taskForWaterfall := waterfallTask{
+			Id:            t.Id,
+			Status:        t.Status,
+			StatusDetails: t.StatusDetails,
+			DisplayName:   t.DisplayName,
+			Activated:     t.Activated,
+			TimeTaken:     t.TimeTaken,
+			StartTime:     t.StartTime.UnixNano(),
+		}
+		taskForWaterfall.Status = uiStatus(taskForWaterfall)
+
+		statusCount.IncrementStatus(taskForWaterfall.Status, taskForWaterfall.StatusDetails)
+
+		waterfallTasks = append(waterfallTasks, taskForWaterfall)
+	}
+	return waterfallTasks, statusCount
+}
+
+// addFailedTests adds all of the failed tests associated with a task to its entry in the WaterfallRow.
+// addFailedAndStartedTests adds all of the failed tests associated with a task to its entry in the waterfallRow
+// and adds the estimated duration to started tasks.
+func addFailedAndStartedTests(waterfallRows map[string]WaterfallRow, failedAndStartedTasks []task.Task) {
+	failedTestsByTaskId := map[string][]string{}
+	expectedDurationByTaskId := map[string]time.Duration{}
+	for _, t := range failedAndStartedTasks {
+		failedTests := []string{}
+		for _, r := range t.LocalTestResults {
+			if r.Status == evergreen.TestFailedStatus {
+				failedTests = append(failedTests, r.TestFile)
+			}
+		}
+		if t.Status == evergreen.TaskStarted {
+			expectedDurationByTaskId[t.Id] = t.ExpectedDuration
+		}
+		failedTestsByTaskId[t.Id] = failedTests
+	}
+	for buildVariant, row := range waterfallRows {
+		for versionId, build := range row.Builds {
+			for i, task := range build.Tasks {
+				if len(failedTestsByTaskId[task.Id]) != 0 {
+					waterfallRows[buildVariant].Builds[versionId].Tasks[i].FailedTestNames = append(
+						waterfallRows[buildVariant].Builds[versionId].Tasks[i].FailedTestNames,
+						failedTestsByTaskId[task.Id]...)
+					sort.Strings(waterfallRows[buildVariant].Builds[versionId].Tasks[i].FailedTestNames)
+				}
+				if duration, ok := expectedDurationByTaskId[task.Id]; ok {
+					waterfallRows[buildVariant].Builds[versionId].Tasks[i].ExpectedDuration = duration
+				}
+			}
+		}
+	}
+}
+
+// Fetch versions until 'numVersionElements' elements are created, including
+// elements consisting of multiple versions rolled-up into one.
+// The skip value indicates how many versions back in time should be skipped
+// before starting to fetch versions, the project indicates which project the
+// returned versions should be a part of.
+func GetWaterfallVersionsAndVariants(
+	skip,
+	numVersionElements int,
+	project *Project,
+	variantQuery string,
+) (versionVariantData, error) {
```

Comment:
public functions that return private data types are incredibly difficult to use and should be avoided.